webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/batch/CleanPromisesFolder.scala

```diff
@@ -0,0 +1,130 @@
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.batch
+
+import better.files.File
+import better.files.File.root
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.domain.logger.ScheduledJobLogger
+import com.normation.rudder.domain.nodes.NodeInfo
+import com.normation.rudder.services.nodes.NodeInfoService
+import monix.execution.Scheduler.{global => scheduler}
+import net.liftweb.common._
+import net.liftweb.util.Helpers.tryo
+
+import scala.concurrent.duration._
+
+
+/**
+ * A naive scheduler which checks every updateInterval if software needs to be deleted
+ */
+class CleanPromisesFolder(
+     nodeInfoService : NodeInfoService
+   , updateInterval  : FiniteDuration
+) {
+
+  val logger = ScheduledJobLogger
+
+
+  if (updateInterval < 1.hour) {
+    logger.info(s"Disable automatic unreferenced promise directory (update interval cannot be less than 1 hour)")
+  } else {
+    logger.debug(s"***** starting batch that purge unreferenced promise directory, every ${updateInterval.toString()} *****")
+    scheduler.scheduleWithFixedDelay(1.hour, updateInterval) {
+
+
+    }
+  }
+
+  def launch = {
+    (for {
+      nodes <- nodeInfoService.getAll()
+      deleted <- cleanPromiseFolderOfDeletedNode(nodes)
+    } yield {
+      deleted
+    }) match {
+      case eb: EmptyBox =>
+        val error = (eb ?~! s"Error when deleting unreferenced promise directory")
+        logger.error(error.messageChain)
+        error.rootExceptionCause.foreach(ex =>
+          logger.error("Exception was:", ex)
+        )
+        error
+      case Full(deleted) =>
+        logger.info(s"Deleted ${deleted.length} unreferenced promise folder")
+        if (logger.isDebugEnabled && deleted.length > 0)
+          logger.debug(s"Deleted promise folder of Nodes: ${deleted.map(_.name).mkString(",")}")
+        Full(deleted)
+    }
+  }
+
+  def cleanPromiseFolderOfDeletedNode(currentNodes : Map[NodeId,NodeInfo]) : Box[Seq[File]] = {
+    import File._
+    def getNodeFolders(file: File): Iterator[File] = {
+      file.children.filter{d => d.name !="rules" && d.isDirectory && d.children.exists{b => b.name == "rules"}}.flatMap {
```

Comment:
java.nio.file.FileSystemException: /var/rudder/share/546f908c-ef75-492d-bc30-5577fa44099a: Too many open files
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:427)
	at java.nio.file.Files.newDirectoryStream(Files.java:457)
	at java.nio.file.Files.list(Files.java:3451)
	at better.files.File.list(File.scala:764)
	at better.files.File.children(File.scala:766)
	at com.normation.rudder.batch.CleanPromisesFolder.$anonfun$cleanPromiseFolderOfDeletedNode$1(CleanPromisesFolder.scala:99)
	at com.normation.rudder.batch.CleanPromisesFolder.$anonfun$cleanPromiseFolderOfDeletedNode$1$adapted(CleanPromisesFolder.scala:99)

Reply:
on my test platform I have 4551 folders in /var/rudder/share.
It means file.children has 4551 open files at least, plus for each of them, we look for the children, so again 4551 files opened.

Reply:
I don't think it should keep them open thought. 
Well, that ticket need perhaps more work to minimize that number.

Reply:
I still have the error when there are folder to delete (if none are to be deleted, all is fine)
[2019-11-15 09:36:21] ERROR scheduledJob - Error when deleting unreferenced promise directory <- /var/rudder/share/be28c259-8e34-4447-a6fe-d661d9af975f: Too many open files
[2019-11-15 09:36:21] ERROR scheduledJob - Exception was:
java.nio.file.FileSystemException: /var/rudder/share/be28c259-8e34-4447-a6fe-d661d9af975f: Too many open files
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:427)
	at java.nio.file.Files.newDirectoryStream(Files.java:457)
	at java.nio.file.Files.list(Files.java:3451)
	at better.files.File.list(File.scala:764)
	at better.files.File.children(File.scala:766)
	at com.normation.rudder.batch.CleanPromisesFolder.$anonfun$cleanPromiseFolderOfDeletedNode$1(CleanPromisesFolder.scala:100)
	at com.normation.rudder.batch.CleanPromisesFolder.$anonfun$cleanPromiseFolderOfDeletedNode$1$adapted(CleanPromisesFolder.scala:100)